distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Table_Helpers.enso

```diff
@@ -211,27 +273,54 @@ select_indices_preserving_order vector indices =
-    if ix < 0 then vector.length+ix else ix
+resolve_index Integer->Integer->Integer
+resolve_index length ix =
+    if ix < 0 then length+ix else ix
 
 ## PRIVATE
    Checks if the given index is in the valid range for the provided vector.
-is_index_valid vector ix =
-    actual_ix = here.resolve_index vector ix
-    actual_ix>=0 && actual_ix<vector.length
+is_index_valid Integer->Integer->Boolean
+is_index_valid length ix =
+    actual_ix = here.resolve_index length ix
+    actual_ix>=0 && actual_ix<length
+
+## PRIVATE
+   Validates a Vector of indices returning a pair of `good_indices` and `problems`
+validate_indices : Integer -> Vector -> Validation_Result Vector Vector
+validate_indices length indices =
+    partitioned_indices = indices.partition (here.is_index_valid length)
+    inbound_indices = partitioned_indices.first
+    oob_indices = partitioned_indices.second
+    oob_problems = if oob_indices.is_empty then [] else
+        [Column_Indexes_Out_Of_Range oob_indices]
+
+    uniques = here.validate_unique inbound_indices v->[Duplicate_Column_Selectors v]
 
-type Repeated_Acc distinct_builder duplicate_builder existing
+    resolver = ix->(here.resolve_index length ix)
+    alias_uniques = here.validate_unique uniques.valid v->[Input_Indices_Already_Matched v] resolver
+    good_indices = alias_uniques.valid.map resolver
+
+    problems = oob_problems + uniques.problems + alias_uniques.problems
+
+    Validation_Result good_indices problems
 
 ## PRIVATE
    Splits a vector into elements which are distinct and the duplicates.
-split_to_distinct_and_duplicates : (Any -> Any) -> Vector -> Pair Vector Vector
-split_to_distinct_and_duplicates vector (on = x->x) =
-    acc = vector.fold (Repeated_Acc Vector.new_builder Vector.new_builder Map.empty) acc-> item->
+   Duplicates are wrapped as an error
+validate_unique : Vector -> (Vector -> Vector) -> (Any -> Any) -> Validation_Result Vector Vector
+validate_unique vector problem_wrapper on=(x->x) =
+    acc = vector.fold [Map.empty, Vector.new_builder, Vector.new_builder] acc-> item->
+        existing = acc.at 0
         key = on item
-        already_present = acc.existing.get_or_else key False
+        already_present = existing.get_or_else key False
         case already_present of
-            True ->
-                Repeated_Acc acc.distinct_builder (acc.duplicate_builder.append item) acc.existing
-            False ->
-                Repeated_Acc (acc.distinct_builder.append item) acc.duplicate_builder (acc.existing.insert key True)
-    Pair acc.distinct_builder.to_vector acc.duplicate_builder.to_vector
+            True -> [existing, acc.at 1, acc.at 2 . append item]
+            False -> [existing.insert key True, acc.at 1 . append item, acc.at 2]
+
+    duplicates = acc.at 2
+    problems = if duplicates.length == 0 then [] else problem_wrapper duplicates.to_vector
+
+    Validation_Result (acc.at 1).to_vector problems
+
+## PRIVATE
+type Validation_Result valid problems
```

Comment:
I think we have a convention where each file should end with a newline.
Should be probably configurable inside the text editor to handle this automatically when saving files.